app/addons/documents/doc-editor/components.react.jsx

```diff
@@ -0,0 +1,540 @@
+        success: this.props.uploadComplete,
+        error: this.error
+      });
+    },
+
+    error: function (resp) {
+      // cancelled uploads throw an ajax error but don't contain a response. We don't want to display an error then anyway
+      if (_.isEmpty(resp.responseText)) {
+        return;
+      }
+      this.setState({
+        errorMessage: 'Could not upload document: ' + JSON.parse(resp.responseText).reason,
+        inProgress: false
+      });
+    },
+
+    uploadProgress: function (percentComplete) {
+      this.setState({
+        loadPercentage: percentComplete
+      });
+    },
+
+    render: function () {
+      var errorClasses = 'alert alert-error';
+      if (this.state.errorMessage === '') {
+        errorClasses += ' hide';
+      }
+      var loadIndicatorClasses = 'progress progress-info';
+      if (!this.state.inProgress) {
+        loadIndicatorClasses += ' hide';
+      }
+
+      return (
+        <div className="modal hide fade upload-file-modal" tabIndex="-1" data-js-visible={this.props.visible}>
+          <div className="modal-header">
+            <button type="button" className="close" onClick={this.closeModal} aria-hidden="true">&times;</button>
+            <h3>Upload Attachment</h3>
+          </div>
+          <div className="modal-body">
+            <div className={errorClasses}>{this.state.errorMessage}</div>
+
+            <div>
+              <form ref="uploadForm" className="form" method="post">
+                <p>
+                  Please select the file you want to upload as an attachment to this document. This creates a new
+                  revision of the document, so it's not necessary to save after uploading.
+                </p>
+                <input ref="attachments" type="file" name="_attachments" />
+                <br />
+              </form>
+
+              <div ref="loadIndicator" className={loadIndicatorClasses}>
+                <div className="bar" style={{ width: this.state.loadPercentage + '%'}}></div>
+              </div>
+            </div>
+
+          </div>
+          <div className="modal-footer">
+             <button href="#" data-bypass="true" className="btn" onClick={this.closeModal}>
+              <i className="icon fonticon-cancel-circled"></i> Cancel
+            </button>
+            <button href="#" id="upload-btn" data-bypass="true" className="btn btn-success save" onClick={this.upload}>
+              <i className="icon fonticon-ok-circled"></i> Upload
+            </button>
+          </div>
+        </div>
+      );
+    }
+  });
+
+
+  var CloneDocModal = React.createClass({
+    propTypes: {
+      visible: React.PropTypes.bool.isRequired
+    },
+
+    getInitialState: function () {
+      return {
+        uuid: null
+      };
+    },
+
+    cloneDoc: function () {
+      if (this.props.onSubmit) {
+        this.props.onSubmit();
+      }
+      Actions.cloneDoc(this.refs.newDocId.getDOMNode().value);
```

Comment:
Urgh, just been looking at this. Can we chat about it tomorrow? Frankly there's downsides to both approaches and I'm not sure a controlled component is an improvement.